crypto/rsa/rsa_ameth.c

```diff
@@ -884,6 +884,111 @@ static int rsa_pss_pkey_import_from(const OSSL_PARAM params[], void *vpctx)
     return rsa_int_import_from(params, vpctx, RSA_FLAG_TYPE_RSASSAPSS);
 }
 
+#define rsa_bn_dup_check(bn) \
+{ \
+  const BIGNUM *f = rsa->bn; \
+  \
+  if (f != NULL && (dupkey->bn = BN_dup(f)) == NULL) \
+      goto err; \
+}
+
+#define pinfo_bn_dup_check(bn) \
+{ \
+  const BIGNUM *f = pinfo->bn; \
+  \
+  if (f != NULL && (duppinfo->bn = BN_dup(f)) == NULL) \
+      goto err; \
+}
+
+static RSA *rsa_dup(const RSA *rsa)
+{
+    RSA *dupkey = NULL;
+    int pnum, i;
+
+    /* Do not try to duplicate foreign RSA keys */
+    if (RSA_get_method(rsa) != RSA_PKCS1_OpenSSL())
+        return NULL;
+
+    if ((dupkey = ossl_rsa_new_with_ctx(rsa->libctx)) == NULL)
+        return NULL;
+
+    /* private and public key */
+    rsa_bn_dup_check(n)
+    rsa_bn_dup_check(e)
+    rsa_bn_dup_check(d)
+
+    /* factors and crt params */
+    rsa_bn_dup_check(p)
+    rsa_bn_dup_check(q)
+    rsa_bn_dup_check(dmp1)
+    rsa_bn_dup_check(dmq1)
+    rsa_bn_dup_check(iqmp)
+
+    /* multiprime */
+    pnum = sk_RSA_PRIME_INFO_num(rsa->prime_infos);
+    if (pnum > 0) {
+        dupkey->prime_infos = sk_RSA_PRIME_INFO_new_reserve(NULL, pnum);
+        for (i = 0; i < pnum; i++) {
+            const RSA_PRIME_INFO *pinfo = NULL;
+            RSA_PRIME_INFO *duppinfo = NULL;
+
+            if ((duppinfo = OPENSSL_zalloc(sizeof(*duppinfo))) == NULL) {
+                ERR_raise(ERR_LIB_RSA, ERR_R_MALLOC_FAILURE);
+                goto err;
+            }
+            /* push first so cleanup in error case works */
+            (void)sk_RSA_PRIME_INFO_push(dupkey->prime_infos, duppinfo);
+
+            pinfo = sk_RSA_PRIME_INFO_value(rsa->prime_infos, i);
+            pinfo_bn_dup_check(r)
+            pinfo_bn_dup_check(d)
+            pinfo_bn_dup_check(t)
+        }
+        if (!ossl_rsa_multip_calc_product(dupkey))
+            goto err;
+    }
+
+    dupkey->pss_params = rsa->pss_params;
+
+    if (rsa->pss != NULL) {
+        dupkey->pss = RSA_PSS_PARAMS_dup(rsa->pss);
+        if (rsa->pss->maskGenAlgorithm != NULL
+            && dupkey->pss->maskGenAlgorithm == NULL) {
+            dupkey->pss->maskHash = ossl_x509_algor_mgf1_decode(rsa->pss->maskGenAlgorithm);
+            if (dupkey->pss->maskHash == NULL)
+                goto err;
+        }
+    }
+
```

Comment:
There a quite a few complex fields in the RSA object.

    int32_t version;
    const RSA_METHOD *meth;
    int flags;
    /* Used to cache montgomery values */
    BN_MONT_CTX *_method_mod_n;
    BN_MONT_CTX *_method_mod_p;
    BN_MONT_CTX *_method_mod_q;
    /*
     * all BIGNUM values are actually in the following data, if it is not
     * NULL
     */
    char *bignum_data;
    BN_BLINDING *blinding;
    BN_BLINDING *mt_blinding;

Reply:
I'll copy the version and flags. The default method is set up by RSA_new and we limit the operation to the default RSA method. The rest is set up automatically when it is needed by operations that use the key so I do not think it needs to be copied.